backend/src/algorithm/index.ts

```diff
@@ -0,0 +1,60 @@
+import { daysPerWeek, maxPeriodsPerDay } from './consts';
+import { GenerateTimetableProps, GenerateTimetableResult, RemainingLectures } from './types';
+import { getTotalRemainingLectures, getTotalTimesPerWeek } from './utils/subject';
+import { setSlot, swapSlots } from './utils/slot';
+import { checkConstraints } from './utils/constraint';
+import { generateTimetableProps } from './seed';
+import { getClassPeriodsPerDay } from './utils/class';
+import { getMatrixHashmap } from './utils/matrix';
+import { shuffleArray } from './utils/array';
+
+export function generateTimetable({ classes, classrooms }: GenerateTimetableProps): GenerateTimetableResult {
+  const teachers = classes
+    .map(class_ => class_.subjects.map(subject => subject.teacher))
+    .flat()
+    .filter((teacher, i, arr) => arr.findIndex(ts => ts.name === teacher.name) === i);
```

Comment:
You assumed correctly. This is the old code I succeeded and I focused more on adding new functionalities and old algorithm improvements than refactoring. Code refactor was planned later on, but I refactored it now. It will take less work later on. 👌